backend/src/game.ts

```diff
@@ -0,0 +1,197 @@
+		const game = games.get(gameId);
+
+		// client-requested game error
+		if (game === undefined) {
+			res.status(404).send({ ok: false , err: `Game ${gameId} not found`});
+			return;
+		}
+
+		// we have no way of validating the host yet
+		// client permission 403 error
+
+		const quiz = game.quizData;
+		// out-of-bounds error
+		if (index >= quiz.questions.length) {
+			res.status(404).send({ ok: false , err: `Question ${index} not found`});
+			return;
+		}
+
+		// start accepting answers for the question index
+		game.activeQuestion = index;
+
+		// show question text and answers on both host and player screens
+		beginQuestion();
+        game.users.set('123', { name: 'bob', answers: [] });
+
+		res.status(200).send({ ok: true });
+	});
+
+	app.post('/games/:gameId/questions/:index/answer', (req, res) => {
+		const gameId: GameId = req.params.gameId;
+		const userId: UserId = req.body.userId;
+		const index = parseInt(req.params.index);
+		const game = games.get(gameId);
+
+		let answer: number;
+        try {
+            answer = parseInt(req.body.answer);
+        } catch (e) {
+            answer = -1;
+        }
+
+		// client-requested game error
+		if (game === undefined) {
+			res.status(404).send({ok: false, err: `Game ${gameId} not found`});
+			return;
+		}
+
+		const quiz = game.quizData;
+
+		// check if question is open
+		if (game.activeQuestion != index) {
+			res.status(400).send({ok: false, err: `Question ${index} is not open for answers`});
+			return;
+		}
+
+		// add answer to userId
+        const user = game.users.get(userId);
+        if (user) {
+            if (user.answers[index] !== undefined) {
+                res.status(400).send({ok: false, err: `Answer for Question ${index} already exists`});
+                return;
+            } else {
+                // if a player has joined late, their previous answers will be undefined
+                user.answers[index] = answer;
+                console.log(user.answers);
+            }
+        } else {
+            res.status(400).send({ok: false, err: `User ${userId} does not exist`});
+        }
+
+        res.status(200).send({ok: true});
+	});
+
+	app.post('/games/:gameId/players', (req, res) => {
+		const body = req.body;
+		const gameId: GameId = req.params.gameId;
+
+		if (!body.username) {
+			res.status(400).send();
+			return;
+		}
+
+		const game = games.get(gameId);
+
+		if (game === undefined) {
+			res.status(404).send();
+			return;
+		}
+
+		// let playerSet = games.get(gameId);
+
+		// if (playerSet === undefined) {
+		// 	playerSet = new Map();
+		// 	games.set(gameId, playerSet);
+		// }
+
+		const username: UserId = body.username;
+		if (game.users.has(username)) {
+			res.status(409).send();
+			return;
+		}
+
+		// Functionally done to make use of code generation
+		// but I don't want to store the id's more than they already are (twice now)
+
+		const id = code.gen(8, [...(game.users.keys(), game.hostId)]);
```

Comment:
See my above comment on `getUsers()` -- the parentheses make this not work but we can just use the function

```suggestion
		const id = code.gen(8, getUsers(game));
```